src/controllers/nodes/nodes_controller.go

```diff
@@ -10,230 +10,192 @@ import (
 
-func (controller *NodesController) onDeletion(node string) error {
-	log.Info("node deletion notification received", "node", node)
+		//Handle unschedulable Nodes, if they have a OneAgent instance
+		if controller.isUnschedulable(&node) {
+			cachedNodeData := CachedNodeInfo{
+				cachedNode: cacheEntry,
+				nodeCache:  nodeCache,
+				nodeName:   nodeName,
+			}
 
-	c, err := controller.getCache()
-	if err != nil {
-		return err
+			if err := controller.markForTermination(dynakube, cachedNodeData); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
 	}
 
-	if err = controller.removeNode(c, node, func(dkName string) (*dynatracev1beta1.DynaKube, error) {
-		var dynaKube dynatracev1beta1.DynaKube
-		if err := controller.client.Get(context.TODO(), client.ObjectKey{Name: dkName, Namespace: controller.namespace}, &dynaKube); err != nil {
-			return nil, err
+	// check node cache for outdated nodes and remove them, to keep cache clean
+	if nodeCache.IsCacheOutdated() {
+		if err := controller.handleOutdatedCache(nodeCache); err != nil {
+			return reconcile.Result{}, err
 		}
-		return &dynaKube, nil
-	}); err != nil {
-		return err
+		nodeCache.UpdateTimestamp()
 	}
-
-	return controller.updateCache(c)
+	return reconcile.Result{}, controller.updateCache(nodeCache, ctx)
 }
 
-func (controller *NodesController) reconcileAll() error {
-	log.Info("reconciling nodes")
-
-	var dkList dynatracev1beta1.DynaKubeList
-	if err := controller.client.List(context.TODO(), &dkList, client.InNamespace(controller.namespace)); err != nil {
+func (controller *NodesController) reconcileNodeDeletion(nodeName string) error {
+	nodeCache, err := controller.getCache()
+	if err != nil {
 		return err
 	}
 
-	dks := make(map[string]*dynatracev1beta1.DynaKube, len(dkList.Items))
-	for i := range dkList.Items {
-		dks[dkList.Items[i].Name] = &dkList.Items[i]
-	}
-
-	c, err := controller.getCache()
+	dynakube, err := controller.determineDynakubeForNode(nodeName)
 	if err != nil {
+		log.Error(err, "error while getting Dynakube for Node on deletion")
 		return err
 	}
 
-	var nodeLst corev1.NodeList
-	if err := controller.client.List(context.TODO(), &nodeLst); err != nil {
+	cachedNodeInfo, err := nodeCache.Get(nodeName)
+	if err != nil {
+		if err == ErrNotFound {
+			// uncached node -> igonoring
+			log.Info("ignoring uncached node on deletion", "node", nodeName)
+			return nil
+		}
+		log.Error(err, "error while getting cachedNode on deletion")
 		return err
 	}
 
-	nodes := map[string]bool{}
-	for i := range nodeLst.Items {
-		node := nodeLst.Items[i]
-		nodes[node.Name] = true
-
-		// Sometimes Azure does not cordon off nodes before deleting them since they use taints,
-		// this case is handled in the update event handler
-		if isUnschedulable(&node) {
-			if err = controller.reconcileUnschedulableNode(&node, c); err != nil {
-				return err
-			}
+	// Node is found in the cluster and in cache, send mark for termination, not found node is handled in err check
```

Comment:
nitpick: a bit confusing comment because it refers to things before and after it,
(the "err check" happens in line 148 
the "node is found logic" starts at 159 
the comment is at 158)